experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,906 @@
+            i3_block = (i3_block & 0x7fff) + (i3 & !7) + (i3 >> 3);
+            i3 = i3 & 7;
+            data_block = ((self.index.get((i3_block + 1) as usize).unwrap() << (2 + (2 * i3)))
+                as u32
+                & 0x30000) as u32;
+            data_block = data_block | self.index.get((i3_block + i3) as usize).unwrap() as u32;
+        }
+        data_block + (c & SMALL_DATA_MASK)
+    }
+
+    /// Internal trie getter for a code point at or above the fast limit. Returns the data index.
+    fn trie_small_index(&self, c: u32) -> u32 {
+        if c >= self.header.high_start {
+            self.header.data_length - HIGH_VALUE_NEG_DATA_OFFSET
+        } else {
+            self.trie_internal_small_index(c) // helper fn
+        }
+    }
+
+    /// Internal trie getter for a code point below the fast limit. Returns the data index.
+    fn trie_fast_index(&self, c: u32) -> u32 {
+        let index_array_pos: u32 = c >> FAST_TYPE_SHIFT;
+        let index_array_val: u16 = self.index.get(index_array_pos as usize).unwrap(); // 1. How to specify type parameter for .index() 2. How to avoid unwrap()?
+        let fast_index_val: u32 = index_array_val as u32 + (c & FAST_TYPE_DATA_MASK);
+        fast_index_val
+    }
+
+    pub fn get(&self, c: u32) -> W {
+        let index: u32 = if c <= T::FAST_MAX {
+            Self::trie_fast_index(self, c)
+        } else if c <= CODE_POINT_MAX {
+            Self::trie_small_index(self, c)
+        } else {
+            self.header.data_length - ERROR_VALUE_NEG_DATA_OFFSET
+        };
+        // need the unwrap because the default value is stored in the data array,
+        // and getting that default value always returns an Option<W>, but need to return W
+        self.data.get(index as usize).unwrap()
+    }
+
+    pub fn get_u32(&self, c: u32) -> u32 {
+        // this is the consistent API that is public-facing for users
+        self.get(c).cast_to_widest()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn check_trie<W: ValueWidth, T: TrieType>(trie: &CodePointTrie<W, T>, check_ranges: &[u32]) {
+        assert_eq!(
+            0,
+            check_ranges.len() % 2,
+            "check_ranges must have an even number of 32-bit values in (limit,value) pairs"
+        );
+
+        let mut i: u32 = 0;
+        let check_range_tuples = check_ranges.chunks(2);
+        // Iterate over each check range
+        for range_tuple in check_range_tuples {
+            let range_end = range_tuple[0];
+            let range_value = range_tuple[1];
+            // Check all values in this range, one-by-one
+            while i < range_end {
+                assert_eq!(range_value, trie.get_u32(i), "trie_get({})", i,);
+                i = i + 1;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod fast_8_test {
+        use super::super::*;
+        use super::check_trie;
+
+        const INDEX: [u16; 1024] = [
```

Comment:
Nit: I find it a bit hard to follow how you've laid out the tests.  It would improve readability if you had a separate file for the binary data for testing, which you import into the test code file.